lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMPass.cpp

```diff
@@ -896,10 +896,24 @@ struct ConvertTritonGPUToLLVM
                          .cast<RankedTensorType>()
                          .getEncoding()
                          .dyn_cast<MfmaEncodingAttr>()) {
-        if (AElType.isBF16() || AElType.isF16() || AElType.isF32() ||
-            AElType.isInteger(8))
+        Type BElType =
+            dotOp.getA().getType().cast<RankedTensorType>().getElementType();
+
+        auto maxBitWidth = std::max(AElType.getIntOrFloatBitWidth(),
+                                    BElType.getIntOrFloatBitWidth());
+
+        // TODO check mfma tensor core version compatibility
+        if (maxBitWidth == 8)
           return;
-        promoteType = builder.getF16Type();
+
+        if (AElType == BElType)
+          return;
+
+        if (maxBitWidth < 16)
+          promoteType = builder.getF16Type();
+
+        if (maxBitWidth < 32)
```

Comment:
Does this mean if data size are different, they are always upcasted to 32? 
and it seems like lines 912 and 913 are useless. Do you understand it correctly?

Reply:
And this case will run `fmuladd` instruction, right?

Reply:
> Does this mean if data size are different, they are always upcasted to 32?

You are right, this is not my intention here.
Interesting how this did not break anything, need to add more asserts.

Reply:
> And this case will run fmuladd instruction, right?

no, we still use MFMA.
Looks like everything works because we do not check assumptions strictly enough...

But I am afraid it could break something in the future, I'll create an issue about this, I think it is a flaw of current design

Reply:
Update:
I forgot about python code, which is doing part of this promotions, so this code is not working in the first place.

But ideally I want to get rid of casts in python during dot operation construction.
This leaks low-level details into triton IR level (not even triton GPU!) and forces us to support two identical logic pieces on python level and in C++.